File: training-data/ruby/url_parser.rb
require 'uri'

def describe(raw)
  u = URI.parse(raw)
  query = u.query ? URI.decode_www_form(u.query).to_h : {}
  puts({ scheme: u.scheme, userinfo: u.userinfo, host: u.host, port: u.port, path: u.path.empty? ? '/' : u.path,
         query: query, fragment: u.fragment }.inspect)
rescue URI::InvalidURIError => e
  puts "invalid: #{e.message}"
end

describe('https://user:pw@example.com:8443/a/b?x=1&y=hello%20world#top')
describe('http://localhost')
describe('http://exa mple.com')
puts URI.join('https://example.com/docs/guide/', '../api?v=2')

File: training-data/csharp/UrlParser.cs
using System;
using System.Linq;
using System.Web;

class UrlParser
{
    static void Describe(string raw)
    {
        if (!Uri.TryCreate(raw, UriKind.Absolute, out var uri))
        {
            Console.WriteLine($"invalid: {raw}");
            return;
        }
        var query = HttpUtility.ParseQueryString(uri.Query);
        string q = string.Join(", ", query.AllKeys.Select(k => $"{k}={query[k]}"));
        Console.WriteLine($"scheme={uri.Scheme} user={uri.UserInfo} host={uri.Host} port={uri.Port} path={uri.AbsolutePath} query=[{q}] fragment={uri.Fragment}");
    }

    static void Main()
    {
        Describe("https://user:pw@example.com:8443/a/b?x=1&y=hello%20world#top");
        Describe("http://localhost");
        Describe("not a url");
        Console.WriteLine(new Uri(new Uri("https://example.com/docs/guide/"), "../api?v=2"));
    }
}
